### ISSUE TITLE
Standardize installation and execution

### ISSUE THREAD

**Issue by mfripp** [switch-model/switch#70]
I think we need to standardize how SWITCH is installed and run, because we seem to have two different versions going on. I'll describe below how I do it, which is my recommended approach. But I would like to hear if people have other suggestions. I would then like to rewrite the instructions so users can get up and running pretty easily.

I install SWITCH by cloning the repository, then cd'ing into the `switch` directory and running `python setup.py develop` or `python setup.py develop --user`. This installs the package in-place, so I can edit it and use it at the same time (and if I want to change to another installation, I just go there and run `python setup.py develop` again). Users who don't want to edit the package can instead run `python setup.py` or `python setup.py --user`. Once I put the package on pypi and conda, they can just do `pip install switch` or `conda install -c switch-model switch` .

These commands work well under anaconda, and would probably work with homebrew python. With the standard system python, users might need to use `sudo` with the system-wide versions.

These commands do a few useful things:

- install the `switch_mod` package in the python system path
- install dependencies specified in setup.py (Pyomo and its dependencies for now, but the list could be extended)
- install the `switch` command-line script (equivalent to `python -m switch_mod.main`)

I then setup models in various locations in my file system (not inside the `switch` folder) and solve them by running `switch solve` or `switch solve-scenarios`. If I edit the local copy of the switch package or use `git pull`, those changes are automatically reflected the next time I use the command-line script. This works well, and makes it easy to setup other users and share models with them (via separate repositories for each model).

As I understand it from the INSTALL file, other SWITCH users are editing their `PYTHONPATH` to point to the switch repository and then using `pip install --user -r pip_requirements.txt` to install the dependencies. I don't know how other users are activating the `switch` command-line script, if they use it at all. This approach has a few disadvantages compared to my approach, which make it difficult to give general installation instructions for new users: 

- it is hard to write general instructions for users to modify their PYTHONPATH appropriately; 
- using pip requires installing pip on some systems, and may cause problems for people who use anaconda
- it's nearly impossible to write cross-platform instructions to install the `switch` command-line script, which is  very handy, especially for new users

All of these problems are addressed automatically by setuptools, which is used by setup.py. This takes care of all the cross-platform issues other than installing the solver.

Any objections to making `setup.py` the standard way to install switch, via the various commands listed in the second paragraph? Or making `switch solve ...` the standard way to run it?

**Comment by rodrigomha** [comment]
Hi everyone,

Mi name is Rodrigo, and I'm working on SWITCH with Benjamin here at the Catholic University in Chile. In my opinion this seems to be the best choice to install and run SWITCH. When I installed SWITCH, I did it using pip and everything else, and it was really troublesome to do it.

Just to let you know, my main problem was to include SWITCH on my Windows PATH. I didn't realize that my PATH has already a length of around 1000 characters and it was not possible to include the switch-master folder on it (I think there is a max of 1024 characters). I fixed it deleting some useless lines on my PATH. 

The main issue is that you don't get a specific error if including the PATH doesn't work (you will know because other things simply won't work), so maybe including a note in the README or setup file about that problem will be a good idea.

Regards,
Rodrigo

**Comment by bmaluenda** [comment]
I agree the setup.py procedure is easier to grasp for new users. We've had a lot of trouble setting up new users in different platforms. It is even convenient if developers switch machines or operating systems, because settling in would be quicker.

The `README` and `INSTALL` files should be modified accordingly with instructions on this setup procedure both for users and developers. Maybe the `how to collaborate.txt` file could be merged in one of those to have a full set of instructions for new developers.

Additionally, the package dependencies in pip_requirements.txt could be added to the setup.py script so that people don't have to manually install them.

It would be great if the installation and usage is standardized. Great initiative.

**Comment by mfripp** [comment]
What do you all think of recommending Anaconda or miniconda as the preferred way to install Python and glpk? With that, the instructions can be pretty streamlined, like this:

INTRODUCTION

This repository contains code and example files for the SWITCH power system planning model. To use this model, you will need to install several things: 

- Python (to run all the code)
- a numerical solver, to do the computations needed to solve the model
- git software, to download SWITCH code and model data from github.com
- Python code for SWITCH and Pyomo (used by SWITCH)

These instructions take you through these steps.

We recommend that you use the Anaconda scientific computing environment to run SWITCH. This provides an easy, cross-platform way to install most of the resources that SWITCH needs, and it avoids interfering with your system's built-in Python installation (if present). The instructions below assume you wiil use the Anaconda distribution. If you prefer to use a different distribution, you will need to adjust the instructions accordingly.

INSTALL ANACONDA AND PYTHON

Download and install the full Anaconda distribution from https://www.continuum.io/downloads or the smaller version from http://conda.pydata.org/miniconda.html . Either way, make sure to select the option to install Python 2.7. (SWITCH is not yet compatible with Python 3.x.).

Note that you do not need administrator privileges to install Anaconda.

INSTALL A SOLVER

You can solve small models, including the SWITCH examples, using the open-source glpk solver. This can be installed as follows:

- Open Terminal.app (OS X) or an Anaconda command prompt (Start -> Anaconda -> Anaconda Prompt)
- Type this command and then press Enter or return:

conda install -c conda-forge glpk

Follow the prompts to install glpk.

To solve larger models, you will need to install the cplex or gurobi solvers, which are an order of magnitude faster than glpk. Both of these have free trials available, and are free long-term for academics. You can install one of these now or after you install SWITCH. More information on these solvers can be found at the following links:

- cplex: https://www.ibm.com/developerworks/community/blogs/jfp/entry/free_cplex_trials?lang=en
- gurobi: http://www.gurobi.com/downloads/download-center

INSTALL GIT SOFTWARE

Open a Terminal window or Anaconda command prompt, as discussed in the previous section, then run this command and follow the prompts:

conda install git

INSTALL SWITCH MODEL CODE AND EXAMPLES

Open a Terminal window or Anaconda command prompt. Then use the 'cd' and 'mkdir' commands to create and/or enter the  directory where you would like to store the SWITCH model code and examples. Once you are in that directory, run the following commands (don't type the comments that start with '#'):

# download SWITCH
git clone https://github.com/switch-model/switch.git

# install the SWITCH model code and required Python packages
cd switch
python setup.py develop --user

This adds the switch_mod package to your Python installation. It also installs the Pyomo Python package, which SWITCH uses to define and solve optimization models. (SWITCH uses Pyomo to define a model, then Pyomo passes the model to the solver and receives the solution, then SWITCH retrieves the solution from Pyomo.)

RUN TESTS

To test your installation, you can run "python run_tests.py".

RUN EXAMPLES

There are numerous example models in subdirectories inside switch/examples. You can run any of these by using the 'cd' command to go into the relevant directory, and then running

switch solve

CONFIGURE YOUR OWN MODELS

At a minimum, each model requires a list of SWITCH modules to define the model and a set of input files to provide the data. The SWITCH framework and individual modules also accept command-line arguments to change their behavior.

Each SWITCH model or collection of models is defined in a specific directory (e.g., examples/3zone_toy). This directory contains one or more subdirectories that hold input data and results (e.g., "inputs" and "outputs"). The models in the examples directory show the type of text files used to provide inputs for a model. You can change any of the model's input data by editing the *.tab files in the input directory.

SWITCH contains a number of different modules, which can be selected and combined to create models with different capabilities and amounts of detail. You can look through the *.py files within switch_mod and its subdirectories to see the standard modules that are available and the columns that each one will read from the input files. You can also add modules of your own by creating Python files in the main model directory and adding their name (without the ".py") to the module list, discussed below. These should define the same functions as the standard modules (e.g., define_components()).

Each model has a text file which lists the modules that will be used for that model. Normally this file is called "modules.txt" and is stored in the main model directory or in an inputs subdirectory. SWITCH will automatically look in those locations for this list; alternatively, you can specify a different file with the "--module-list" argument.

Use "switch --help", "switch solve --help" or  "switch solve-scenarios --help" to see a list of command-line arguments that are available. 

You can specify these arguments on the command line when you solve the model (e.g., "switch solve --solver cplex"). You can also place frequently used arguments in a file called "options.txt" in the main model directory. These can all be on one line, or they can be placed on multiple lines for easier readability (be sure to include the "--" part in all the argument names in options.txt). The "switch solve" command first reads all the arguments from options.txt, and then applies any arguments you specified on the command line. If the same argument is specified multiple times, the last one takes priority. 

You can also define scenarios, which are sets of command-line arguments to define different models. These additional arguments can be placed in a scenario list file, usually called "scenarios.txt" in the main model directory (or you can use a different file specified by "--scenario-list"). Each scenario should be defined on a single line, which includes a "--scenario-name" argument and any other arguments needed to define the scenario. "switch solve-scenarios" will solve all the scenarios listed in this file. For each scenario, it will first apply all arguments from options.txt, then arguments from the relevant line of scenarios.txt, then any arguments specified on the command line.

After the model runs, results will be written in tab-separated text files (with extension .tsv or .tab) in the "outputs" directory (or some other directory specified via the "--outputs-dir" argument).

UPDATE TO LATEST VERSION OF SWITCH

You can pull the latest version of the SWITCH code and examples from github.com at any time by launching a Terminal window or Anaconda prompt, then cd'ing into the 'switch' directory and running this command:

git pull

This will attempt to merge your local changes with changes in the main repository. If there are any conflicts, you can follow the instructions given by the git command to resolve them.

**Comment by bmaluenda** [comment]
Anaconda has been the easiest way for me to setup SWITCH for new users, so I'm all for it.

I would add some notes on what the `setup.py` script actually does, so user-devs have an easier time setting up multiple Switch packages.

Josiah was considering adding some tweaks to the setup.py script, so I would wait for his comments before posting the new instructions.

PS: I think we need to get a website up.

**Comment by josiahjohnston** [comment]
I really like the idea of streamlining installation and usage. Thank you for putting this together.

See pull request #74 for my attempt at implementation. I had issues with using setup.py directly, so I wrote a draft INSTALL to use setup.py in conjunction with pip based on the [Python Packaging User Guide](https://packaging.python.org/distributing/). See details below.

Using switch as the main entry point sounds great, but for me I couldn't get this script if I included the --user flag. This is true whether I use `python setup.py develop --user` or `pip install --editable --user .` I haven't found a workaround for that, and have only tested it on a mac so far.

Directing python newbies to Anaconda seems fine, and veteran python developers can use whatever setup they like. In either case, we should be able to assume that pip is available, since [pip is included in Anaconda and Miniconda](http://conda.pydata.org/docs/using/pkgs.html) and pip is a core tool of python developers. 

We definitely need setuptools and setup.py since they are needed for pip, but I'm concerned about setup.py being the main entry point because all of the python packaging advice I read recommends using pip. For non-developers, distributing on PyPi / conda should be the end-game, I think. For developers, running `pip install --editable .` instead of `python setup.py develop` should be recommended, possibly using virtual environments. pip will parse setup.py, but does a better job at managing dependencies, and allows you to uninstall the package if desired (unlike setup.py by itself). Pip also has support for "extras" like the R or psql interfaces (unlike setup.py by itself)

When I first tried to use setup.py on my computer, it totally mangled all of my python dependencies in a way that I had never seen before, and it didn't give me a clean way to undo its mess. I ended up removing homebrew and anaconda, upgrading my OS to Sierra to get an os-distribution of python that was semi-modern, and trying again from a relatively clean slate. I was able to get it working after that, but the experience makes me cautious about recommending setup.py to developers with complex setups. If everyone else can test setup.py 

I like the details in the readme file you wrote. Could we use it for INSTALL.txt rather than README? One comment is to mention [Cbc](https://projects.coin-or.org/Cbc) in the list of solvers; it's open source and a lot faster than glpk.

References I found useful:
* [Python Packaging User Guide: Packaging and Distributing Projects](https://packaging.python.org/distributing/)
* [install_requires vs Requirements files](https://packaging.python.org/requirements/#install-requires-vs-requirements-files)

**Comment by mfripp** [comment]
I agree, we should make `pip install --editable .` the standard installation method (or eventually `conda` for non-developer users). I was already leaning that way, and this seals the deal. I just posted another pull request involving some changes to setup.py, and I'll go back and revise that soon.

When using setup.py directly, I think `python setup.py develop --user` is useful for HPC systems, where it installs the command-line scripts somewhere in the user path, and doesn't require system-level access. It may also be useful on Unixy systems running the system-installed Python. But the '--user' flag is unnecessary with the conda python, and I think it may install the scripts in some location under `~/.local` that is not in the standard search path. If we move to pip, this may not be an issue. Or it may be. I'll see if I can come up with some simple instructions that cover the normal range of cases.

I had a cbc installation in my readme.txt at one point. With conda, you just need to run `conda install -c conda-forge coincbc`. However, no one provides a conda package for a Windows build of coincbc, and coincbc solved my test case slower than glpk, so I removed it to keep things simple. I can add it back as an optional step, with notes.
